internal/acs/runner.go

```diff
@@ -0,0 +1,237 @@
+		level.Debug(collectorParams.Logger).Log("Cleaning up old collectors...")
+		fmt.Println("TOTAL TIME NEEDED::::::", time.Since(startTime))
+		for key, value := range collectorParams.Collector.MetricsData {
+			indexCache[key] = value
+		}
+		level.Debug(collectorParams.Logger).Log("Collection completed.")
+		level.Debug(collectorParams.Logger).Log("Cleaning up old collectors...")
+		staleData := cleanupStaleData(indexCache, collectorParams.Collector)
+		level.Info(collectorParams.Logger).Log("msg", "Cleanup compeleted", "stale records", staleData)
+	}
+}
+
+func cleanupStaleData(indexCache map[string]prometheus.Metric, collector *collector.CVECollector) int {
+	counter := 0
+	for index := range collector.MetricsData {
+		if _, exists := indexCache[index]; !exists {
+			delete(collector.MetricsData, index)
+			counter++
+		}
+	}
+	return counter
+}
+func publishMetricData(client acsclient.Client, collector *collector.CVECollector, workerPoolSize int, logger log.Logger) error {
+	clusters, err := client.GetClusters()
+	if err != nil {
+		return fmt.Errorf("error getting clusters:%v", err)
+	}
+	workerPool := worker.NewWorkerPool(workerPoolSize, true)
+	for _, cluster := range clusters {
+		workerPool.Run(func() error {
+			return processClusterCVEs(cluster, client, collector, logger)
+		})
+	}
+	if err := workerPool.Wait(); err != nil {
+		return fmt.Errorf("errors in worker pools:: %v", err)
+	}
+	return nil
+}
+
+func processClusterCVEs(cluster acsclient.Cluster, client acsclient.Client, collector *collector.CVECollector, logger log.Logger) error {
+	query := acsclient.Query{
+		"Cluster": cluster.Name,
+	}
+	exportDeployments, err := client.GetExportDeployments(query)
+	if err != nil {
+		return fmt.Errorf("errors getting export deployments:: %v", err)
+	}
+	exportImages, err := client.GetExportImages(query)
+	if err != nil {
+		return fmt.Errorf("errors getting export images:: %v", err)
+	}
+	return processNamespaces(query, client, exportDeployments, exportImages, collector, logger)
+}
+func processNamespaces(query acsclient.Query, client acsclient.Client, exportDeployments []acsclient.ExportDeployment, exportImages []acsclient.ExportImage, collector *collector.CVECollector, logger log.Logger) error {
+	namespaces, err := client.GetNamespaces(query)
+	if err != nil {
+		return fmt.Errorf("errors getting namespaces:: %v", err)
+	}
+	for _, namespace := range namespaces {
+		query["Namespace"] = namespace.Metadata.Name
+		cveMap := summarizeCVEData(query, exportDeployments, exportImages)
+		publishCVEmetrics(cveMap, collector, logger)
+	}
+	return nil
+}
+
+func summarizeCVEData(query acsclient.Query, exportDeployments []acsclient.ExportDeployment, exportImages []acsclient.ExportImage) map[string]cveSummaryData {
+	cveMap := map[string]cveSummaryData{}
+	for _, deployment := range exportDeployments {
+		if deployment.Namespace == query["Namespace"] {
+			populateCveMap(query, deployment, cveMap, exportImages)
+		}
+	}
```

Comment:
@hemslo  yes exportDeployments is full list of deployments in a particular cluster hence it will be one per deployment. Its not just used to compare namespace but also to count the no. of deployments in which the image is present

Reply:
deployment count is combine image info with deployment info, here we just need lookup by namespace, construct lookup map to speedup search